narwhal/consensus/src/tests/randomized_tests.rs

```diff
@@ -0,0 +1,568 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+use crate::bullshark::Bullshark;
+use crate::consensus::ConsensusProtocol;
+use crate::consensus::ConsensusState;
+use crate::consensus_utils::make_consensus_store;
+use crate::metrics::ConsensusMetrics;
+use config::{Committee, Stake};
+use crypto::PublicKey;
+use fastcrypto::hash::Hash;
+use fastcrypto::hash::HashFunction;
+use prometheus::Registry;
+use rand::distributions::Bernoulli;
+use rand::distributions::Distribution;
+use rand::prelude::SliceRandom;
+use rand::rngs::StdRng;
+use rand::Rng;
+use rand::SeedableRng;
+use std::collections::{BTreeSet, HashMap, HashSet, VecDeque};
+use std::num::NonZeroUsize;
+use std::ops::RangeInclusive;
+use std::sync::Arc;
+use test_utils::mock_certificate_with_rand;
+use test_utils::CommitteeFixture;
+#[allow(unused_imports)]
+use tokio::sync::mpsc::channel;
+use types::Round;
+use types::{Certificate, CertificateDigest};
+
+#[derive(Copy, Clone, Debug)]
+pub struct FailureModes {
+    // The probability of having failures per round. The failures should
+    // be <=f , otherwise no DAG could be created. The provided number gives the probability of having
+    // failures up to f. Ex for input `failures_probability = 0.2` it means we'll have 20% change of
+    // having failures up to 33% of the nodes.
+    pub nodes_failure_probability: f64,
+
+    // The percentage of slow nodes we want to introduce to our sample. Basically a slow node is one
+    // that might be able to produce certificates, but those are never get referenced by others. That has
+    // as an effect that when they are leaders might also not get enough support - or no support at all.
+    // For example, a value of 0.2 means that we want up to 20% of our nodes to behave as slow nodes.
+    pub slow_nodes_percentage: f64,
+
+    // The probability of failing to include a slow node certificate to from the certificates of next
+    // round. For example a value of 0.1 means that 10% of the time fail get referenced by the
+    // certificates of the next round.
+    pub slow_nodes_failure_probability: f64,
+}
+
+struct ExecutionPlan {
+    certificates: Vec<Certificate>,
+}
+
+impl ExecutionPlan {
+    fn hash(&self) -> [u8; crypto::DIGEST_LENGTH] {
+        let mut hasher = crypto::DefaultHashFunction::new();
+        self.certificates.iter().for_each(|c| {
+            hasher.update(c.digest());
+        });
+        hasher.finalize().into()
+    }
+}
+
+#[tokio::test]
+#[ignore]
+async fn bullshark_randomised_tests() {
+    // Configuration regarding the randomized tests. The tests will run for different values
+    // on the below parameters to increase the different cases we can generate.
+
+    // A range of gc_depth to be used
+    const GC_DEPTH: RangeInclusive<Round> = 4..=15;
```

Comment:
How long does it take to test with more realistic numbers, e.g. GC depth of 50, 100 nodes, 100 ~ 1000 rounds?

Reply:
That's a good point , I haven't tested with such large numbers, but to be honest I was thinking that whatever we can achieve with experiments of ~20 rounds we could pretty much achieve with even higher ones. So based on my understanding edge cases could be caught without having the need to run for large round of numbers. I can totally check though.

Reply:
I would say that most pathogening cases will appear when we do not commit for a long time and then we suddenly commit but different leaders per node. So gc is of little importance since it only works when we do commit